### PR TITLE
Implement WebGL getUniform API

### PR DESCRIPTION
**Pull request by gootorov**
Reference: [WebGL 1](https://www.khronos.org/registry/webgl/specs/latest/1.0/#5.14.10)

---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [ ] These changes fix #__ (github issue number if applicable).

- [x] There are tests for these changes

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/20330)

### CODE REVIEW

**Comment by gootorov** [comment]
Don't merge this yet. This is work in progress and we need to wait for https://github.com/servo/gleam/pull/147 to land first.

**Comment by CYBAI** [comment]
```
{"status": "FAIL", "group": "default", "message": "assert_true: getError expected: INVALID_OPERATION. Was NO_ERROR : after evaluating: contextA.shaderSource(shaderB, 'foo') expected true got false", "stack": "reportTestResultsToHarness/<@http://web-platform.test:8000/_mozilla/webgl/conformance-1.0.3/resources/js-test-pre.js:88:9\nTest.prototype.step@http://web-platform.test:8000/resources/testharness.js:1494:20\ntest@http://web-platform.test:8000/resources/testharness.js:511:9\nreportTestResultsToHarness@http://web-platform.test:8000/_mozilla/webgl/conformance-1.0.3/resources/js-test-pre.js:87:7\ntestFailed@http://web-platform.test:8000/_mozilla/webgl/conformance-1.0.3/resources/js-test-pre.js:159:5\nglErrorShouldBe@http://web-platform.test:8000/_mozilla/webgl/conformance-1.0.3/conformance/resources/webgl-test-utils.js:1406:5\nshouldGenerateGLError@http://web-platform.test:8000/_mozilla/webgl/conformance-1.0.3/conformance/resources/webgl-test-utils.js:1357:12\n@http://web-platform.test:8000/_mozilla/webgl/conformance-1.0.3/conformance/context/incorrect-context-object-behaviour.html:70:1\n", "subtest": "WebGL test #8: getError expected: INVALID_OPERATION. Was NO_ERROR : after evaluating: contextA.shaderSource(shaderB, 'foo')", "test": "/_mozilla/webgl/conformance-1.0.3/conformance/context/incorrect-context-object-behaviour.html", "line": 183119, "action": "test_result", "expected": "PASS"}
{"status": "FAIL", "group": "default", "message": "assert_true: getError expected: INVALID_OPERATION. Was NO_ERROR : after evaluating: contextA.bindAttribLocation(programB, 0, 'foo') expected true got false", "stack": "reportTestResultsToHarness/<@http://web-platform.test:8000/_mozilla/webgl/conformance-1.0.3/resources/js-test-pre.js:88:9\nTest.prototype.step@http://web-platform.test:8000/resources/testharness.js:1494:20\ntest@http://web-platform.test:8000/resources/testharness.js:511:9\nreportTestResultsToHarness@http://web-platform.test:8000/_mozilla/webgl/conformance-1.0.3/resources/js-test-pre.js:87:7\ntestFailed@http://web-platform.test:8000/_mozilla/webgl/conformance-1.0.3/resources/js-test-pre.js:159:5\nglErrorShouldBe@http://web-platform.test:8000/_mozilla/webgl/conformance-1.0.3/conformance/resources/webgl-test-utils.js:1406:5\nshouldGenerateGLError@http://web-platform.test:8000/_mozilla/webgl/conformance-1.0.3/conformance/resources/webgl-test-utils.js:1357:12\n@http://web-platform.test:8000/_mozilla/webgl/conformance-1.0.3/conformance/context/incorrect-context-object-behaviour.html:71:1\n", "subtest": "WebGL test #9: getError expected: INVALID_OPERATION. Was NO_ERROR : after evaluating: contextA.bindAttribLocation(programB, 0, 'foo')", "test": "/_mozilla/webgl/conformance-1.0.3/conformance/context/incorrect-context-object-behaviour.html", "line": 183120, "action": "test_result", "expected": "PASS"}
{"status": "FAIL", "group": "default", "message": "assert_true: getError expected: INVALID_OPERATION. Was NO_ERROR : after evaluating: contextA.bindFramebuffer(contextA.FRAMEBUFFER, frameBufferB) expected true got false", "stack": "reportTestResultsToHarness/<@http://web-platform.test:8000/_mozilla/webgl/conformance-1.0.3/resources/js-test-pre.js:88:9\nTest.prototype.step@http://web-platform.test:8000/resources/testharness.js:1494:20\ntest@http://web-platform.test:8000/resources/testharness.js:511:9\nreportTestResultsToHarness@http://web-platform.test:8000/_mozilla/webgl/conformance-1.0.3/resources/js-test-pre.js:87:7\ntestFailed@http://web-platform.test:8000/_mozilla/webgl/conformance-1.0.3/resources/js-test-pre.js:159:5\nglErrorShouldBe@http://web-platform.test:8000/_mozilla/webgl/conformance-1.0.3/conformance/resources/webgl-test-utils.js:1406:5\nshouldGenerateGLError@http://web-platform.test:8000/_mozilla/webgl/conformance-1.0.3/conformance/resources/webgl-test-utils.js:1357:12\n@http://web-platform.test:8000/_mozilla/webgl/conformance-1.0.3/conformance/context/incorrect-context-object-behaviour.html:72:1\n", "subtest": "WebGL test #10: getError expected: INVALID_OPERATION. Was NO_ERROR : after evaluating: contextA.bindFramebuffer(contextA.FRAMEBUFFER, frameBufferB)", "test": "/_mozilla/webgl/conformance-1.0.3/conformance/context/incorrect-context-object-behaviour.html", "line": 183121, "action": "test_result", "expected": "PASS"}
{"status": "FAIL", "group": "default", "message": "assert_true: getError expected: INVALID_OPERATION. Was NO_ERROR : after evaluating: contextA.bindRenderbuffer(contextA.RENDERBUFFER, renderBufferB) expected true got false", "stack": "reportTestResultsToHarness/<@http://web-platform.test:8000/_mozilla/webgl/conformance-1.0.3/resources/js-test-pre.js:88:9\nTest.prototype.step@http://web-platform.test:8000/resources/testharness.js:1494:20\ntest@http://web-platform.test:8000/resources/testharness.js:511:9\nreportTestResultsToHarness@http://web-platform.test:8000/_mozilla/webgl/conformance-1.0.3/resources/js-test-pre.js:87:7\ntestFailed@http://web-platform.test:8000/_mozilla/webgl/conformance-1.0.3/resources/js-test-pre.js:159:5\nglErrorShouldBe@http://web-platform.test:8000/_mozilla/webgl/conformance-1.0.3/conformance/resources/webgl-test-utils.js:1406:5\nshouldGenerateGLError@http://web-platform.test:8000/_mozilla/webgl/conformance-1.0.3/conformance/resources/webgl-test-utils.js:1357:12\n@http://web-platform.test:8000/_mozilla/webgl/conformance-1.0.3/conformance/context/incorrect-context-object-behaviour.html:73:1\n", "subtest": "WebGL test #11: getError expected: INVALID_OPERATION. Was NO_ERROR : after evaluating: contextA.bindRenderbuffer(contextA.RENDERBUFFER, renderBufferB)", "test": "/_mozilla/webgl/conformance-1.0.3/conformance/context/incorrect-context-object-behaviour.html", "line": 183122, "action": "test_result", "expected": "PASS"}
{"status": "FAIL", "group": "default", "message": "assert_true: getError expected: INVALID_OPERATION. Was NO_ERROR : after evaluating: contextA.bindTexture(contextA.TEXTURE_2D, textureB) expected true got false", "stack": "reportTestResultsToHarness/<@http://web-platform.test:8000/_mozilla/webgl/conformance-1.0.3/resources/js-test-pre.js:88:9\nTest.prototype.step@http://web-platform.test:8000/resources/testharness.js:1494:20\ntest@http://web-platform.test:8000/resources/testharness.js:511:9\nreportTestResultsToHarness@http://web-platform.test:8000/_mozilla/webgl/conformance-1.0.3/resources/js-test-pre.js:87:7\ntestFailed@http://web-platform.test:8000/_mozilla/webgl/conformance-1.0.3/resources/js-test-pre.js:159:5\nglErrorShouldBe@http://web-platform.test:8000/_mozilla/webgl/conformance-1.0.3/conformance/resources/webgl-test-utils.js:1406:5\nshouldGenerateGLError@http://web-platform.test:8000/_mozilla/webgl/conformance-1.0.3/conformance/resources/webgl-test-utils.js:1357:12\n@http://web-platform.test:8000/_mozilla/webgl/conformance-1.0.3/conformance/context/incorrect-context-object-behaviour.html:74:1\n", "subtest": "WebGL test #12: getError expected: INVALID_OPERATION. Was NO_ERROR : after evaluating: contextA.bindTexture(contextA.TEXTURE_2D, textureB)", "test": "/_mozilla/webgl/conformance-1.0.3/conformance/context/incorrect-context-object-behaviour.html", "line": 183123, "action": "test_result", "expected": "PASS"}
{"status": "FAIL", "group": "default", "message": "assert_true: getError expected: INVALID_OPERATION. Was NO_ERROR : after evaluating: contextA.framebufferRenderbuffer(contextA.FRAMEBUFFER, contextA.DEPTH_ATTACHMENT, contextA.RENDERBUFFER, renderBufferB) expected true got false", "stack": "reportTestResultsToHarness/<@http://web-platform.test:8000/_mozilla/webgl/conformance-1.0.3/resources/js-test-pre.js:88:9\nTest.prototype.step@http://web-platform.test:8000/resources/testharness.js:1494:20\ntest@http://web-platform.test:8000/resources/testharness.js:511:9\nreportTestResultsToHarness@http://web-platform.test:8000/_mozilla/webgl/conformance-1.0.3/resources/js-test-pre.js:87:7\ntestFailed@http://web-platform.test:8000/_mozilla/webgl/conformance-1.0.3/resources/js-test-pre.js:159:5\nglErrorShouldBe@http://web-platform.test:8000/_mozilla/webgl/conformance-1.0.3/conformance/resources/webgl-test-utils.js:1406:5\nshouldGenerateGLError@http://web-platform.test:8000/_mozilla/webgl/conformance-1.0.3/conformance/resources/webgl-test-utils.js:1357:12\n@http://web-platform.test:8000/_mozilla/webgl/conformance-1.0.3/conformance/context/incorrect-context-object-behaviour.html:75:1\n", "subtest": "WebGL test #13: getError expected: INVALID_OPERATION. Was NO_ERROR : after evaluating: contextA.framebufferRenderbuffer(contextA.FRAMEBUFFER, contextA.DEPTH_ATTACHMENT, contextA.RENDERBUFFER, renderBufferB)", "test": "/_mozilla/webgl/conformance-1.0.3/conformance/context/incorrect-context-object-behaviour.html", "line": 183124, "action": "test_result", "expected": "PASS"}
{"status": "FAIL", "group": "default", "message": "assert_true: getError expected: INVALID_OPERATION. Was NO_ERROR : after evaluating: contextA.framebufferTexture2D(contextA.FRAMEBUFFER, contextA.COLOR_ATTACHMENT0, contextA.TEXTURE_2D, textureB, 0) expected true got false", "stack": "reportTestResultsToHarness/<@http://web-platform.test:8000/_mozilla/webgl/conformance-1.0.3/resources/js-test-pre.js:88:9\nTest.prototype.step@http://web-platform.test:8000/resources/testharness.js:1494:20\ntest@http://web-platform.test:8000/resources/testharness.js:511:9\nreportTestResultsToHarness@http://web-platform.test:8000/_mozilla/webgl/conformance-1.0.3/resources/js-test-pre.js:87:7\ntestFailed@http://web-platform.test:8000/_mozilla/webgl/conformance-1.0.3/resources/js-test-pre.js:159:5\nglErrorShouldBe@http://web-platform.test:8000/_mozilla/webgl/conformance-1.0.3/conformance/resources/webgl-test-utils.js:1406:5\nshouldGenerateGLError@http://web-platform.test:8000/_mozilla/webgl/conformance-1.0.3/conformance/resources/webgl-test-utils.js:1357:12\n@http://web-platform.test:8000/_mozilla/webgl/conformance-1.0.3/conformance/context/incorrect-context-object-behaviour.html:76:1\n", "subtest": "WebGL test #14: getError expected: INVALID_OPERATION. Was NO_ERROR : after evaluating: contextA.framebufferTexture2D(contextA.FRAMEBUFFER, contextA.COLOR_ATTACHMENT0, contextA.TEXTURE_2D, textureB, 0)", "test": "/_mozilla/webgl/conformance-1.0.3/conformance/context/incorrect-context-object-behaviour.html", "line": 183125, "action": "test_result", "expected": "PASS"}
{"status": "FAIL", "group": "default", "message": "assert_true: getError expected: INVALID_OPERATION. Was INVALID_ENUM : after evaluating: contextA.getProgramParameter(programB, 0) expected true got false", "stack": "reportTestResultsToHarness/<@http://web-platform.test:8000/_mozilla/webgl/conformance-1.0.3/resources/js-test-pre.js:88:9\nTest.prototype.step@http://web-platform.test:8000/resources/testharness.js:1494:20\ntest@http://web-platform.test:8000/resources/testharness.js:511:9\nreportTestResultsToHarness@http://web-platform.test:8000/_mozilla/webgl/conformance-1.0.3/resources/js-test-pre.js:87:7\ntestFailed@http://web-platform.test:8000/_mozilla/webgl/conformance-1.0.3/resources/js-test-pre.js:159:5\nglErrorShouldBe@http://web-platform.test:8000/_mozilla/webgl/conformance-1.0.3/conformance/resources/webgl-test-utils.js:1406:5\nshouldGenerateGLError@http://web-platform.test:8000/_mozilla/webgl/conformance-1.0.3/conformance/resources/webgl-test-utils.js:1357:12\n@http://web-platform.test:8000/_mozilla/webgl/conformance-1.0.3/conformance/context/incorrect-context-object-behaviour.html:77:1\n", "subtest": "WebGL test #15: getError expected: INVALID_OPERATION. Was INVALID_ENUM : after evaluating: contextA.getProgramParameter(programB, 0)", "test": "/_mozilla/webgl/conformance-1.0.3/conformance/context/incorrect-context-object-behaviour.html", "line": 183126, "action": "test_result", "expected": "PASS"}
{"status": "FAIL", "group": "default", "message": "assert_true: getError expected: INVALID_OPERATION. Was NO_ERROR : after evaluating: contextA.getProgramInfoLog(programB, 0) expected true got false", "stack": "reportTestResultsToHarness/<@http://web-platform.test:8000/_mozilla/webgl/conformance-1.0.3/resources/js-test-pre.js:88:9\nTest.prototype.step@http://web-platform.test:8000/resources/testharness.js:1494:20\ntest@http://web-platform.test:8000/resources/testharness.js:511:9\nreportTestResultsToHarness@http://web-platform.test:8000/_mozilla/webgl/conformance-1.0.3/resources/js-test-pre.js:87:7\ntestFailed@http://web-platform.test:8000/_mozilla/webgl/conformance-1.0.3/resources/js-test-pre.js:159:5\nglErrorShouldBe@http://web-platform.test:8000/_mozilla/webgl/conformance-1.0.3/conformance/resources/webgl-test-utils.js:1406:5\nshouldGenerateGLError@http://web-platform.test:8000/_mozilla/webgl/conformance-1.0.3/conformance/resources/webgl-test-utils.js:1357:12\n@http://web-platform.test:8000/_mozilla/webgl/conformance-1.0.3/conformance/context/incorrect-context-object-behaviour.html:78:1\n", "subtest": "WebGL test #16: getError expected: INVALID_OPERATION. Was NO_ERROR : after evaluating: contextA.getProgramInfoLog(programB, 0)", "test": "/_mozilla/webgl/conformance-1.0.3/conformance/context/incorrect-context-object-behaviour.html", "line": 183127, "action": "test_result", "expected": "PASS"}
{"status": "FAIL", "group": "default", "message": "assert_true: getError expected: INVALID_OPERATION. Was INVALID_ENUM : after evaluating: contextA.getShaderParameter(shaderB, 0) expected true got false", "stack": "reportTestResultsToHarness/<@http://web-platform.test:8000/_mozilla/webgl/conformance-1.0.3/resources/js-test-pre.js:88:9\nTest.prototype.step@http://web-platform.test:8000/resources/testharness.js:1494:20\ntest@http://web-platform.test:8000/resources/testharness.js:511:9\nreportTestResultsToHarness@http://web-platform.test:8000/_mozilla/webgl/conformance-1.0.3/resources/js-test-pre.js:87:7\ntestFailed@http://web-platform.test:8000/_mozilla/webgl/conformance-1.0.3/resources/js-test-pre.js:159:5\nglErrorShouldBe@http://web-platform.test:8000/_mozilla/webgl/conformance-1.0.3/conformance/resources/webgl-test-utils.js:1406:5\nshouldGenerateGLError@http://web-platform.test:8000/_mozilla/webgl/conformance-1.0.3/conformance/resources/webgl-test-utils.js:1357:12\n@http://web-platform.test:8000/_mozilla/webgl/conformance-1.0.3/conformance/context/incorrect-context-object-behaviour.html:79:1\n", "subtest": "WebGL test #17: getError expected: INVALID_OPERATION. Was INVALID_ENUM : after evaluating: contextA.getShaderParameter(shaderB, 0)", "test": "/_mozilla/webgl/conformance-1.0.3/conformance/context/incorrect-context-object-behaviour.html", "line": 183128, "action": "test_result", "expected": "PASS"}
{"status": "FAIL", "group": "default", "message": "assert_true: getError expected: INVALID_OPERATION. Was NO_ERROR : after evaluating: contextA.getShaderInfoLog(shaderB, 0) expected true got false", "stack": "reportTestResultsToHarness/<@http://web-platform.test:8000/_mozilla/webgl/conformance-1.0.3/resources/js-test-pre.js:88:9\nTest.prototype.step@http://web-platform.test:8000/resources/testharness.js:1494:20\ntest@http://web-platform.test:8000/resources/testharness.js:511:9\nreportTestResultsToHarness@http://web-platform.test:8000/_mozilla/webgl/conformance-1.0.3/resources/js-test-pre.js:87:7\ntestFailed@http://web-platform.test:8000/_mozilla/webgl/conformance-1.0.3/resources/js-test-pre.js:159:5\nglErrorShouldBe@http://web-platform.test:8000/_mozilla/webgl/conformance-1.0.3/conformance/resources/webgl-test-utils.js:1406:5\nshouldGenerateGLError@http://web-platform.test:8000/_mozilla/webgl/conformance-1.0.3/conformance/resources/webgl-test-utils.js:1357:12\n@http://web-platform.test:8000/_mozilla/webgl/conformance-1.0.3/conformance/context/incorrect-context-object-behaviour.html:80:1\n", "subtest": "WebGL test #18: getError expected: INVALID_OPERATION. Was NO_ERROR : after evaluating: contextA.getShaderInfoLog(shaderB, 0)", "test": "/_mozilla/webgl/conformance-1.0.3/conformance/context/incorrect-context-object-behaviour.html", "line": 183129, "action": "test_result", "expected": "PASS"}
{"status": "FAIL", "group": "default", "message": "assert_true: getError expected: INVALID_OPERATION. Was NO_ERROR : after evaluating: contextA.getShaderSource(shaderB) expected true got false", "stack": "reportTestResultsToHarness/<@http://web-platform.test:8000/_mozilla/webgl/conformance-1.0.3/resources/js-test-pre.js:88:9\nTest.prototype.step@http://web-platform.test:8000/resources/testharness.js:1494:20\ntest@http://web-platform.test:8000/resources/testharness.js:511:9\nreportTestResultsToHarness@http://web-platform.test:8000/_mozilla/webgl/conformance-1.0.3/resources/js-test-pre.js:87:7\ntestFailed@http://web-platform.test:8000/_mozilla/webgl/conformance-1.0.3/resources/js-test-pre.js:159:5\nglErrorShouldBe@http://web-platform.test:8000/_mozilla/webgl/conformance-1.0.3/conformance/resources/webgl-test-utils.js:1406:5\nshouldGenerateGLError@http://web-platform.test:8000/_mozilla/webgl/conformance-1.0.3/conformance/resources/webgl-test-utils.js:1357:12\n@http://web-platform.test:8000/_mozilla/webgl/conformance-1.0.3/conformance/context/incorrect-context-object-behaviour.html:81:1\n", "subtest": "WebGL test #19: getError expected: INVALID_OPERATION. Was NO_ERROR : after evaluating: contextA.getShaderSource(shaderB)", "test": "/_mozilla/webgl/conformance-1.0.3/conformance/context/incorrect-context-object-behaviour.html", "line": 183130, "action": "test_result", "expected": "PASS"}
{"status": "FAIL", "group": "default", "message": "assert_true: getError expected: INVALID_OPERATION. Was NO_ERROR : after evaluating: contextA.getUniformLocation(programB, 'u_modelViewProjMatrix') expected true got false", "stack": "reportTestResultsToHarness/<@http://web-platform.test:8000/_mozilla/webgl/conformance-1.0.3/resources/js-test-pre.js:88:9\nTest.prototype.step@http://web-platform.test:8000/resources/testharness.js:1494:20\ntest@http://web-platform.test:8000/resources/testharness.js:511:9\nreportTestResultsToHarness@http://web-platform.test:8000/_mozilla/webgl/conformance-1.0.3/resources/js-test-pre.js:87:7\ntestFailed@http://web-platform.test:8000/_mozilla/webgl/conformance-1.0.3/resources/js-test-pre.js:159:5\nglErrorShouldBe@http://web-platform.test:8000/_mozilla/webgl/conformance-1.0.3/conformance/resources/webgl-test-utils.js:1406:5\nshouldGenerateGLError@http://web-platform.test:8000/_mozilla/webgl/conformance-1.0.3/conformance/resources/webgl-test-utils.js:1357:12\n@http://web-platform.test:8000/_mozilla/webgl/conformance-1.0.3/conformance/context/incorrect-context-object-behaviour.html:83:1\n", "subtest": "WebGL test #21: getError expected: INVALID_OPERATION. Was NO_ERROR : after evaluating: contextA.getUniformLocation(programB, 'u_modelViewProjMatrix')", "test": "/_mozilla/webgl/conformance-1.0.3/conformance/context/incorrect-context-object-behaviour.html", "line": 183132, "action": "test_result", "expected": "PASS"}
```

File: components/canvas/webgl_thread.rs [hunk 1]
```diff
@@ -1179,6 +1181,56 @@ impl WebGLImpl {
         chan.send(location).unwrap();
     }
 
+    #[allow(unsafe_code)]
```
Comment (nox) [inline components/canvas/webgl_thread.rs]: Why allow unsafe code here?

Comment (gootorov) [reply components/canvas/webgl_thread.rs]: No reason to. Probably added it accidentally, thanks for noticing!

**Comment by nox** [comment]
@gootorov Do you need help with something for that PR?

**Comment by gootorov** [comment]
@nox, It's going well, thanks! :) We only need to wait for https://github.com/servo/gleam/pull/159 fix to land and this PR would be almost ready. Locally, everything works and tests pass with these changes.

**Comment by gootorov** [comment]
@nox, It's going well, thanks! :) We only need to wait for https://github.com/servo/gleam/pull/159 fix to land and this PR would be almost ready. Locally, everything works and tests pass with these changes.

**Comment by gootorov** [comment]
This PR is ready, except that `WebGLActiveInfo::Type()` API always returns an array type, which causes the following tests to fail:
https://github.com/servo/servo/blob/6b7957d9a009af5e5b5d36df70f100a5b5b64917/tests/wpt/mozilla/tests/webgl/conformance-1.0.3/conformance/uniforms/uniform-location.html#L73-L75
Instead of receiving a single uniform, we get an array `[333,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]`.


We need to fix `WebGLActiveInfo::Type()` for these changes to work correctly.

**Comment by jdm** [comment]
@avadacatavra Review ping.

**Comment by jdm** [comment]
This PR will need to be rebased first before it can merge. I'll make that happen.

**Comment by jdm** [comment]
@bors-servo try
@avadacatavra I had to make a bunch of adjustments to the patch as part of the rebase. Could you take another look at the changes?

**Comment by avadacatavra** [comment]
@bors-servo r+

@jdm i think it still looks good

**Comment by gootorov** [comment]
Looks like the previous failure was some infrastructure error.

@bors-servo try

**Comment by jdm** [comment]
Superseded by the changes in #21202. Sorry there were so many issues here!

**Pull request closed without merging**